faucet/src/faucet.rs

```diff
@@ -283,52 +284,63 @@ pub fn run_local_faucet(
     run_local_faucet_with_port(faucet_keypair, sender, per_time_cap, 0)
 }
 
-pub fn run_faucet(
+pub async fn run_faucet(
     faucet: Arc<Mutex<Faucet>>,
     faucet_addr: SocketAddr,
     send_addr: Option<Sender<SocketAddr>>,
 ) {
-    let socket = TcpListener::bind(&faucet_addr).unwrap();
+    let listener = TcpListener::bind(&faucet_addr).await.unwrap();
     if let Some(send_addr) = send_addr {
-        send_addr.send(socket.local_addr().unwrap()).unwrap();
+        send_addr.send(listener.local_addr().unwrap()).unwrap();
     }
     info!("Faucet started. Listening on: {}", faucet_addr);
     info!(
         "Faucet account address: {}",
         faucet.lock().unwrap().faucet_keypair.pubkey()
     );
 
-    let done = socket
-        .incoming()
-        .map_err(|e| debug!("failed to accept socket; error = {:?}", e))
-        .for_each(move |socket| {
-            let faucet2 = faucet.clone();
-            let framed = BytesCodec::new().framed(socket);
-            let (writer, reader) = framed.split();
+    loop {
+        let _faucet = faucet.clone();
+        match listener.accept().await {
+            Ok((stream, _)) => {
+                tokio::spawn(async move {
+                    if let Err(e) = process(stream, _faucet).await {
+                        info!("failed to process request; error = {:?}", e);
+                    }
+                });
+            }
+            Err(e) => debug!("failed to accept socket; error = {:?}", e),
+        }
+    }
+}
 
-            let processor = reader.and_then(move |bytes| {
-                match faucet2.lock().unwrap().process_faucet_request(&bytes) {
+async fn process(
+    stream: TokioTcpStream,
+    faucet: Arc<Mutex<Faucet>>,
+) -> Result<(), Box<dyn std::error::Error>> {
+    let framed = BytesCodec::new().framed(stream);
```

Comment:
Unrelated to these changes, but I wonder if we have an unbounded read here, like we did in ip-echo-server (#13568)

Reply:
Looks like. Thanks for pointing out that PR!
Shouldn't be an issue to do an read_exact here. I'll plug that in shortly.

Reply:
@t-nelson can you please take a peek at https://github.com/solana-labs/solana/pull/14336/commits/fbd0645ff179647c63acb6e4788322ab8177d1ae?